Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp

```diff
@@ -56,16 +59,66 @@ void HTMLScriptElement::set_non_blocking(Badge<HTMLDocumentParser>, bool non_blo
 
 void HTMLScriptElement::execute_script()
 {
-    document().run_javascript(m_script_source);
+    if (m_preparation_time_document.ptr() != &document()) {
+        dbgln("HTMLScriptElement: Refusing to run script because the preparation time document is not the same as the node document.");
+        return;
+    }
+
+    if (m_script_source.is_null()) {
+        dbgln("HTMLScriptElement: Refusing to run script because the script source is null.");
+        dispatch_event(DOM::Event::create(HTML::EventNames::error));
+        return;
+    }
+
+    bool incremented_destructive_writes_counter = false;
 
-    if (has_attribute(HTML::AttributeNames::src))
-        dispatch_event(DOM::Event::create(EventNames::load));
+    if (m_from_an_external_file || m_is_module) {
+        document().increment_ignore_destructive_writes_counter();
+        incremented_destructive_writes_counter = true;
+    }
+
+    if (!m_is_module) {
+        auto old_current_script = document().current_script();
+        if (!is<DOM::ShadowRoot>(root()))
+            document().set_current_script({}, this);
+        else
+            document().set_current_script({}, nullptr);
+
+        if (m_from_an_external_file)
+            dbgln<HTML_SCRIPT_DEBUG>("HTMLScriptElement: Running script {}", attribute(HTML::AttributeNames::src));
+        else
+            dbgln<HTML_SCRIPT_DEBUG>("HTMLScriptElement: Running inline script");
+
+        document().run_javascript(m_script_source);
+
+        document().set_current_script({}, old_current_script);
+    } else {
+        ASSERT(!document().current_script());
+        TODO();
+    }
+
+    if (incremented_destructive_writes_counter)
+        document().decrement_ignore_destructive_writes_counter();
+
+    if (m_from_an_external_file)
+        dispatch_event(DOM::Event::create(HTML::EventNames::load));
+}
+
+// https://mimesniff.spec.whatwg.org/#javascript-mime-type-essence-match
+static bool is_javascript_mime_type_essence_match(const String& mime_type)
+{
+    auto lowercase_mime_type = mime_type.to_lowercase();
+    return lowercase_mime_type.is_one_of("application/ecmascript", "application/javascript", "application/x-ecmascript", "application/x-javascript", "text/ecmascript", "text/javascript", "text/javascript1.0", "text/javascript1.1", "text/javascript1.2", "text/javascript1.3", "text/javascript1.4", "text/javascript1.5", "text/jscript", "text/livescript", "text/x-ecmascript", "text/x-javascript");
 }
```

Comment:
This is not entirely correct, the "essence" of a mime type is:

> The essence of a MIME type mimeType is mimeType’s type, followed by U+002F (/), followed by mimeType’s subtype.

I.e. the parameters are stripped from the mime type - maybe worth a FIXME comment. Eventually we should move mime types from plain strings to an AK helper class like `URL` IMO (I did some work around this a while ago but never finished it), it's a lot nicer than operating on raw strings :)

Reply:
Yeah I agree, especially since there's a _[whole spec](https://mimesniff.spec.whatwg.org/)_ for it.